MekHQ/unittests/mekhq/campaign/parts/equipment/AmmoBinTest.java

```diff
@@ -20,13 +20,1433 @@
+        DocumentBuilder db = MekHqXmlUtil.newSafeDocumentBuilder();
+
+        // Parse using builder to get DOM representation of the XML file
+        Document xmlDoc = db.parse(new ByteArrayInputStream(xml.getBytes()));
+
+        Element partElt = xmlDoc.getDocumentElement();
+        assertEquals("part", partElt.getNodeName());
+
+        // Deserialize the AmmoBin
+        Part deserializedPart = Part.generateInstanceFromXML(partElt, new Version("1.0.0"));
+        assertNotNull(deserializedPart);
+        assertTrue(deserializedPart instanceof AmmoBin);
+
+        AmmoBin deserialized = (AmmoBin) deserializedPart;
+
+        // Check that we deserialized the part correctly.
+        assertEquals(ammoBin.getId(), deserialized.getId());
+        assertEquals(ammoBin.getEquipmentNum(), deserialized.getEquipmentNum());
+        assertEquals(ammoBin.getType(), deserialized.getType());
+        assertEquals(ammoBin.getShotsNeeded(), deserialized.getShotsNeeded());
+        assertEquals(ammoBin.getName(), deserialized.getName());
+    }
+
+    @Test
+    public void changeMunitionTest() {
+        AmmoType isSRM2Ammo = getAmmoType("ISSRM2 Ammo");
+        Campaign mockCampaign = mock(Campaign.class);
+
+        int equipmentNum = 19;
+        AmmoBin ammoBin = new AmmoBin(0, isSRM2Ammo, equipmentNum, 0, false, false, mockCampaign);
+
+        Unit mockUnit = mock(Unit.class);
+        Entity mockEntity = mock(Entity.class);
+        when(mockUnit.getEntity()).thenReturn(mockEntity);
+        Mounted mockMounted = mock(Mounted.class);
+        when(mockMounted.getType()).thenReturn(isSRM2Ammo);
+        when(mockMounted.getBaseShotsLeft()).thenReturn(isSRM2Ammo.getShots());
+        when(mockEntity.getEquipment(eq(equipmentNum))).thenReturn(mockMounted);
+        doAnswer(invocation -> {
+            // Update the ammo type returned by mounted
+            AmmoType newAmmoType = invocation.getArgument(0);
+            when(mockMounted.getType()).thenReturn(newAmmoType);
+            return null;
+        }).when(mockMounted).changeAmmoType(any());
+        doAnswer(invocation -> {
+            // Update the shots left when we're updated
+            int shotsLeft = invocation.getArgument(0);
+            when(mockMounted.getBaseShotsLeft()).thenReturn(shotsLeft);
+            return null;
+        }).when(mockMounted).setShotsLeft(anyInt());
+        ammoBin.setUnit(mockUnit);
+
+        // Before we do anything there should be nothing to fix on a full bin.
+        assertFalse(ammoBin.needsFixing());
+
+        // Pick a different munition type
+        AmmoType isSRM2InfernoAmmo = getAmmoType("ISSRM2 Inferno Ammo");
+        ammoBin.changeMunition(isSRM2InfernoAmmo);
+
+        assertEquals(isSRM2InfernoAmmo, ammoBin.getType());
+        assertTrue(ammoBin.needsFixing());
+
+        // TODO: determine if this should be uncommented
+        // assertEquals(isSRM2InfernoAmmo.getShots(), ammoBin.getShotsNeeded());
```

Comment:
?

Reply:
The expected behavior was a bit up in the air when I wrote that.